### PR TITLE
stabilize TenantControl test one more time

### LINKED ISSUE #17990: Multiple TenantControlTest failures [HZ-1843]
- Possible commit `Adds tenant control (commit: 02b2b45) (detail)`
- Fails on `Hazelcast-4.master-Solaris-OracleJDK8`
- Fails on [Build #255 (Dec 19, 2020 7:20:46 PM)](http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-4.master-Solaris-OracleJDK8/com.hazelcast$hazelcast/255/#showFailuresLink)
- Sample Error
```
expected:<1> but was:<0>
```
- Sample Stacktrace
```
java.lang.AssertionError: expected:<1> but was:<0>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at com.hazelcast.spi.tenantcontrol.TenantControlTest.basicSetTest(TenantControlTest.java:197)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```
- Kindly check

### PR DESCRIPTION
**Pull request by lprimak**
fixes #17990

### CODE REVIEW

**Comment by lprimak** [comment]
@vbekiaris wanna merge this one :)?

**Comment by vbekiaris** [comment]
hey @lprimak , thanks for the PR!

Based on the comment above the assertions:
```
        // expecting tenant context is created & closed 4 times:
        // CachePutOperation
        // AbstractCacheRecordStore.getExpiryPolicy
        // CacheGetOperation
        // CachePutOperation
```

Context create & close operations occur synchronously in those ops, so I don't expect adding eventual assertion will fix the test: by the time those `get`s/`put`s are done the context create/close are also done. What do you think?

**Comment by lprimak** [comment]
Hi @vbekiaris,

I ran repeated tests before and after, and 'before' fails 1/20 times, and the 'after' doesn't
I think it has something to do with the instance getting shut down before the ops complete.
It's been a while since I ran these tests and don't have further details (that's all I remember)

File: hazelcast/src/test/java/com/hazelcast/spi/tenantcontrol/TenantControlTest.java [hunk 1]
```diff
@@ -142,9 +142,9 @@ public void testTenantControl_executionBeforeAfterOps() {
         // these operations don't require tenant control:
         // AddCacheConfigOperation
         // TenantControlReplicationOperation
-        assertEquals(4, setTenantCount.get());
-        assertEquals(4, closeTenantCount.get());
-        assertEquals(1, registerTenantCount.get());
+        assertEqualsEventually(4, setTenantCount);
```
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/spi/tenantcontrol/TenantControlTest.java:145]:
```suggestion
        assertEquals(4, setTenantCount);
```

File: hazelcast/src/test/java/com/hazelcast/spi/tenantcontrol/TenantControlTest.java [hunk 2]
```diff
@@ -142,9 +142,9 @@ public void testTenantControl_executionBeforeAfterOps() {
...
+        assertEqualsEventually(4, closeTenantCount);
+        assertEqualsEventually(1, registerTenantCount);
```
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/spi/tenantcontrol/TenantControlTest.java:147]:
```suggestion
        assertEquals(1, registerTenantCount);
```

**Review by vbekiaris — APPROVED** [review]
Eventual assertion is required only for `closeTenantCount`.
Reasoning: operations notify responses before executing
their afterRun & closing the tenant context, so it is possible
their futures are completed (so get/put are done) before
tenant context is closed.

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
thanks for the fix @lprimak . I pushed commits to only perform eventual assertion on `closeTenantCount`. As soon as build is green we'll merge this.

**Pull request merged**